File: collector/raptor_collector/named_conf.py
"""Parse named.conf zone statements and includes."""

import os
import re
from dataclasses import dataclass
from typing import Iterable, List, Optional, Set


@dataclass(frozen=True)
class BindZone:
    name: str
    file_path: Optional[str]
    zone_type: str = ""


_ZONE_RE = re.compile(
    r'zone\s+"([^"]+)"\s*(?:in\s+)?\{(.*?)\};',
    re.IGNORECASE | re.DOTALL,
)
_FILE_RE = re.compile(r'\bfile\s+"([^"]+)"', re.IGNORECASE)
_TYPE_RE = re.compile(r'\btype\s+(\w+)', re.IGNORECASE)
_INCLUDE_RE = re.compile(r'\binclude\s+"([^"]+)"\s*;', re.IGNORECASE)


def _strip_named_comments(text: str) -> str:
    text = re.sub(r"/\*.*?\*/", " ", text, flags=re.DOTALL)
    text = re.sub(r"//.*?$", " ", text, flags=re.MULTILINE)
    text = re.sub(r"#.*?$", " ", text, flags=re.MULTILINE)
    return text


def parse_named_conf(text: str, conf_dir: str = "") -> List[BindZone]:
    cleaned = _strip_named_comments(text)
    zones: List[BindZone] = []
    for match in _ZONE_RE.finditer(cleaned):
        name = match.group(1).strip().rstrip(".").lower()
        body = match.group(2)
        file_match = _FILE_RE.search(body)
        type_match = _TYPE_RE.search(body)
        file_path = file_match.group(1) if file_match else None
        if file_path and not os.path.isabs(file_path) and conf_dir:
            file_path = os.path.normpath(os.path.join(conf_dir, file_path))
        zones.append(
            BindZone(
                name=name,
                file_path=file_path,
                zone_type=(type_match.group(1).lower() if type_match else ""),
            )
        )
    return zones


def iter_named_conf_files(path: str, seen: Optional[Set[str]] = None) -> Iterable[str]:
    seen = seen if seen is not None else set()
    resolved = os.path.abspath(path)
    if resolved in seen or not os.path.isfile(resolved):
        return
    seen.add(resolved)
    yield resolved
    directory = os.path.dirname(resolved)
    try:
        text = open(resolved, "r", encoding="utf-8", errors="replace").read()
    except OSError:
        return
    for match in _INCLUDE_RE.finditer(_strip_named_comments(text)):
        included = match.group(1)
        if not os.path.isabs(included):
            included = os.path.join(directory, included)
        yield from iter_named_conf_files(included, seen)


def load_bind_zones(conf_path: str) -> List[BindZone]:
    zones: List[BindZone] = []
    for path in iter_named_conf_files(conf_path):
        directory = os.path.dirname(path)
        try:
            text = open(path, "r", encoding="utf-8", errors="replace").read()
        except OSError:
            continue
        zones.extend(parse_named_conf(text, directory))
    return zones

File: collector/internal/namedconf/namedconf.go
package namedconf

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

type BindZone struct {
	Name     string
	FilePath string
	Type     string
}

var (
	zoneRe    = regexp.MustCompile(`(?is)zone\s+"([^"]+)"\s*(?:in\s+)?\{(.*?)\};`)
	fileRe    = regexp.MustCompile(`(?i)\bfile\s+"([^"]+)"`)
	typeRe    = regexp.MustCompile(`(?i)\btype\s+(\w+)`)
	includeRe = regexp.MustCompile(`(?i)\binclude\s+"([^"]+)"\s*;`)
	blockCmt  = regexp.MustCompile(`(?s)/\*.*?\*/`)
	lineCmt1  = regexp.MustCompile(`(?m)//.*$`)
	lineCmt2  = regexp.MustCompile(`(?m)#.*$`)
)

func stripComments(text string) string {
	text = blockCmt.ReplaceAllString(text, " ")
	text = lineCmt1.ReplaceAllString(text, " ")
	return lineCmt2.ReplaceAllString(text, " ")
}

func Parse(text, confDir string) []BindZone {
	cleaned := stripComments(text)
	var zones []BindZone
	for _, match := range zoneRe.FindAllStringSubmatch(cleaned, -1) {
		name := strings.ToLower(strings.TrimSuffix(strings.TrimSpace(match[1]), "."))
		body := match[2]
		filePath := ""
		if fm := fileRe.FindStringSubmatch(body); fm != nil {
			filePath = fm[1]
			if !filepath.IsAbs(filePath) && confDir != "" {
				filePath = filepath.Clean(filepath.Join(confDir, filePath))
			}
		}
		ztype := ""
		if tm := typeRe.FindStringSubmatch(body); tm != nil {
			ztype = strings.ToLower(tm[1])
		}
		zones = append(zones, BindZone{Name: name, FilePath: filePath, Type: ztype})
	}
	return zones
}

func Load(confPath string) []BindZone {
	seen := map[string]bool{}
	var zones []BindZone
	var walk func(string)
	walk = func(path string) {
		abs, err := filepath.Abs(path)
		if err != nil || seen[abs] {
			return
		}
		seen[abs] = true
		raw, err := os.ReadFile(abs)
		if err != nil {
			return
		}
		dir := filepath.Dir(abs)
		zones = append(zones, Parse(string(raw), dir)...)
		for _, inc := range includeRe.FindAllStringSubmatch(stripComments(string(raw)), -1) {
			included := inc[1]
			if !filepath.IsAbs(included) {
				included = filepath.Join(dir, included)
			}
			walk(included)
		}
	}
	walk(confPath)
	return zones
}
